mimesis/builtins/pl.py

```diff
@@ -0,0 +1,100 @@
+"""Specific data provider for Poland (pl)."""
+
+from calendar import isleap
+from typing import Optional
+
+from mimesis.builtins.base import BaseSpecProvider
+from mimesis.enums import Gender
+from mimesis.typing import DateTime
+
+__all__ = ['PolandSpecProvider']
+
+
+class PolandSpecProvider(BaseSpecProvider):
+    """Class that provides special data for Poland (pl)."""
+
+    class Meta:
+        """The name of the provider."""
+
+        name = 'poland_provider'
+
+    def nip(self) -> str:
+        """Generate random valid 10-digit NIP.
+
+        :return: valid 10-digit NIP
+        """
+        nip_digits = list(map(int, str(self.random.randint(101, 998))))
+        for _ in range(6):
+            nip_digits.append(self.random.randint(0, 9))
+        args = (6, 5, 7, 2, 3, 4, 5, 6, 7)
+        sum_v = sum(map(lambda x: x[0] * x[1], zip(args, nip_digits)))
+        checksum_digit = sum_v % 11
+        while checksum_digit > 9:
+            checksum_digit = 0
+        nip_digits.append(checksum_digit)
+        return ''.join(str(d) for d in nip_digits)
+
+    def pesel(self, birth_date: Optional[DateTime]=None,
+              gender: Optional[Gender]=None) -> str:
+        """Generate random 11-digit PESEL.
+
+        :param birth_date: initial birth date (optional)
+        :param gender: gender of person
+        :return: valid 11-digit PESEL
+        """
+        if not birth_date:
+            year = self.random.randint(1800, 2299) % 100
+            month = self.random.randint(1, 12)
+            if month in (4, 6, 9, 10):
+                day = self.random.randint(1, 30)
+            elif month == 2:
+                if isleap(year):
+                    day = self.random.randint(1, 29)
+                else:
+                    day = self.random.randint(1, 28)
+            else:
+                day = self.random.randint(1, 31)
+        else:
+            year = birth_date.date().year
+            month = birth_date.date().month
+        pesel_digits = list(map(int, '{:02d}'.format(year)))
+        if 1800 <= year <= 1899:
+            month += 80
+        elif 1900 <= year <= 1999:
```

Comment:
Why just do we check it?

Reply:
This checks are needed due to the way polish PESEL identification number is created. Here a bit of explanation from (Wiki)[https://en.wikipedia.org/wiki/PESEL]:
>The PESEL system has been designed to cover five centuries. To distinguish people born in     different centuries, numbers are added to the MM field: 
>  * for birthdates between 1900 and 1999 - no change to MM field is made (see below)
>  * for other birthdates:
>      * 1800-1899 - month field number is increased by 80
>       * 2000-2099 - month + 20
>      * 2100-2199 - month + 40
>       * 2200-2299 - month + 60
>For example, a person born on December 24, 2002 would have a PESEL number starting with 023224 and person born on December 24, 1902 would have a PESEL number starting with 021224.